File: src/main/java/com/demonwav/mcdev/platform/mixin/insight/ShadowLineMarkerProvider.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mixin.insight;

import com.demonwav.mcdev.asset.MixinAssets;
import com.demonwav.mcdev.platform.mixin.util.MixinUtils;
import com.demonwav.mcdev.platform.mixin.util.ShadowedMembers;
import com.demonwav.mcdev.util.McEditorUtil;

import com.intellij.codeHighlighting.Pass;
import com.intellij.codeInsight.TargetElementUtil;
import com.intellij.codeInsight.daemon.GutterIconNavigationHandler;
import com.intellij.codeInsight.daemon.LineMarkerInfo;
import com.intellij.codeInsight.daemon.LineMarkerProviderDescriptor;
import com.intellij.codeInsight.daemon.MergeableLineMarkerInfo;
import com.intellij.featureStatistics.FeatureUsageTracker;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.markup.GutterIconRenderer;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiIdentifier;
import com.intellij.psi.PsiNameIdentifierOwner;
import com.intellij.util.Function;
import com.intellij.util.NullableFunction;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.List;

import javax.swing.Icon;

public class ShadowLineMarkerProvider extends LineMarkerProviderDescriptor {
    @Nullable
    @Override
    public LineMarkerInfo getLineMarkerInfo(@NotNull PsiElement element) {
        final ShadowedMembers shadowedMembers = MixinUtils.getShadowedElement(element);
        if (shadowedMembers.getTargets().isEmpty()) {
            return null;
        }

        if (!(element instanceof PsiNameIdentifierOwner)) {
            return null;
        }

        final PsiIdentifier identifier = (PsiIdentifier) ((PsiNameIdentifierOwner) element).getNameIdentifier();
        if (identifier == null) {
            return null;
        }

        return new ShadowLineMarkerInfo(
            identifier,
            identifier.getTextRange(),
            Pass.UPDATE_ALL,
            getIcon(),
            (mouseEvent, psiElement) -> {
                final Editor editor = FileEditorManager.getInstance(element.getProject()).getSelectedTextEditor();
                if (editor == null) {
                    return;
                }

                // TODO create popup and choose out of the list
                FeatureUsageTracker.getInstance().triggerFeatureUsed("navigation.goto.declaration");
                PsiElement navElement = shadowedMembers.getTargets().get(0).getNavigationElement();
                navElement = TargetElementUtil.getInstance().getGotoDeclarationTarget(shadowedMembers.getTargets().get(0), navElement);
                if (navElement != null) {
                    McEditorUtil.gotoTargetElement(navElement, editor, element.getContainingFile());
                }
            }
        );
    }

    @Override
    public void collectSlowLineMarkers(@NotNull List<PsiElement> elements, @NotNull Collection<LineMarkerInfo> result) {

    }

    @Nullable
    @Override
    public String getName() {
        return "Shadow line marker";
    }

    @NotNull
    @Override
    public Icon getIcon() {
        return MixinAssets.SHADOW;
    }

    private static final class ShadowLineMarkerInfo extends MergeableLineMarkerInfo<PsiElement> {
        ShadowLineMarkerInfo(@NotNull PsiElement element,
                             @NotNull TextRange textRange,
                             int updatePass,
                             @NotNull Icon icon,
                             @Nullable GutterIconNavigationHandler<PsiElement> navHandler) {
            super(
                element,
                textRange,
                icon,
                updatePass,
                (NullableFunction<PsiElement, String>) element1 -> "Go to Shadow element",
                navHandler,
                GutterIconRenderer.Alignment.LEFT
            );
        }

        @Override
        public boolean canMergeWith(@NotNull MergeableLineMarkerInfo<?> info) {
            return info instanceof ShadowLineMarkerInfo;
        }

        @Override
        public Icon getCommonIcon(@NotNull List<MergeableLineMarkerInfo> infos) {
            return MixinAssets.SHADOW;
        }

        @NotNull
        @Override
        public Function<? super PsiElement, String> getCommonTooltip(@NotNull List<MergeableLineMarkerInfo> infos) {
            return element -> "Multiple Shadows";
        }

        @Override
        public GutterIconRenderer.Alignment getCommonIconAlignment(@NotNull List<MergeableLineMarkerInfo> infos) {
            return GutterIconRenderer.Alignment.LEFT;
        }
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/mixin/insight/ShadowLineMarkerProvider.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mixin.insight

import com.demonwav.mcdev.asset.MixinAssets
import com.demonwav.mcdev.platform.mixin.util.findFirstShadowTarget
import com.demonwav.mcdev.platform.mixin.util.findShadowTargets
import com.demonwav.mcdev.util.ifEmpty
import com.intellij.codeInsight.daemon.GutterIconNavigationHandler
import com.intellij.codeInsight.daemon.LineMarkerInfo
import com.intellij.codeInsight.daemon.LineMarkerProviderDescriptor
import com.intellij.codeInsight.daemon.MergeableLineMarkerInfo
import com.intellij.codeInsight.daemon.impl.PsiElementListNavigator
import com.intellij.ide.util.DefaultPsiElementCellRenderer
import com.intellij.ide.util.MethodCellRenderer
import com.intellij.openapi.editor.markup.GutterIconRenderer
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiField
import com.intellij.psi.PsiIdentifier
import com.intellij.psi.PsiMember
import com.intellij.psi.PsiMethod
import com.intellij.util.FunctionUtil
import java.awt.event.MouseEvent
import javax.swing.Icon

class ShadowLineMarkerProvider : LineMarkerProviderDescriptor(), GutterIconNavigationHandler<PsiIdentifier> {

    companion object {
        private val ICON: Icon = MixinAssets.SHADOW
        private val TOOLTIP_FUNCTION = FunctionUtil.constant<Any, String>("Go to Shadow element")
    }

    override fun getName() = "@Shadow line marker"
    override fun getIcon() = ICON

    override fun getLineMarkerInfo(element: PsiElement): LineMarkerInfo<PsiIdentifier>? {
        if (element !is PsiMember) {
            return null
        }

        val identifier = when (element) {
            is PsiMethod -> element.nameIdentifier
            is PsiField -> element.nameIdentifier
            else -> null
        } ?: return null

        // Check if @Shadow actually has a target
        element.findFirstShadowTarget() ?: return null
        return LineMarker(identifier, this)
    }

    override fun collectSlowLineMarkers(elements: List<PsiElement>, result: Collection<LineMarkerInfo<PsiElement>>) {
    }

    override fun navigate(e: MouseEvent, elt: PsiIdentifier) {
        val member = elt.parent as? PsiMember ?: return
        val targets = member.findShadowTargets().ifEmpty { return }

        // Create correct renderer for element type
        val renderer = when (member) {
            is PsiMethod -> MethodCellRenderer(true)
            is PsiField -> DefaultPsiElementCellRenderer()
            else -> return
        }

        PsiElementListNavigator.openTargets(
            e, targets.toTypedArray(),
            "Choose target class of ${member.name!!}", null, renderer
        )
    }

    private class LineMarker(
        identifier: PsiIdentifier,
        navHandler: GutterIconNavigationHandler<PsiIdentifier>
    ) : MergeableLineMarkerInfo<PsiIdentifier>(
        identifier,
        identifier.textRange,
        ICON,
        TOOLTIP_FUNCTION,
        navHandler,
        GutterIconRenderer.Alignment.LEFT
    ) {

        override fun canMergeWith(info: MergeableLineMarkerInfo<*>) = info is LineMarker
        override fun getCommonTooltip(infos: List<MergeableLineMarkerInfo<PsiElement>>) = TOOLTIP_FUNCTION
        override fun getCommonIcon(infos: List<MergeableLineMarkerInfo<PsiElement>>) = ICON
    }
}
